### PR TITLE
 [FLINK-9964][table] Add a full RFC-compliant CSV table format factory

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This PR adds a RFC-4180 standard compliant CSV table format. 

## Brief change log

It adds:

- `CsvRowSerializationSchema`
- `CsvRowDeserializationSchema`
- A new `CSV` descriptor
- `CsvRowTableFormatFactory`
- documentation and tests

This PR further improves #6541 to make the CSV format robust and configurable enough to serve most use cases. It integrates nicely with most SQL types.

It deprecates the "old CSV" descriptor stack and prepares also for FLINK-7050 (#4660).

The old CSV descriptor is still available under as "OldCsv".

## Verifying this change

Tests are added for descriptor, factory, and de/serialization schema.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Comment by dawidwys** [comment]
@flinkbot approve description
@flinkbot approve consensus

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvRowDeSerializationSchemaTest.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+			deserializationSchema.deserialize("Test,null,Test".getBytes()); // null not supported
+			fail("Missing field should cause failure.");
+		} catch (IOException e) {
+			// valid exception
+		}
+
+		deserializationSchema.setIgnoreParseErrors(true);
+
+		// unsupported null for integer
+		assertEquals(Row.of("Test", null, "Test"), deserializationSchema.deserialize("Test,null,Test".getBytes()));
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvRowDeSerializationSchemaTest.java]: nit: Could we have those as separate tests? Personally prefer this approach as it immediately shows what failed (by the test name).

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+			return convertObjectArray(node, ((ObjectArrayTypeInfo) info).getComponentInfo());
+		} else if (info instanceof PrimitiveArrayTypeInfo &&
+				((PrimitiveArrayTypeInfo) info).getComponentType() == Types.BYTE) {
+			return convertByteArray((TextNode) node);
+		} else {
+			throw new RuntimeException("Unsupported type information '" + info + "' for node: " + node);
+		}
+	}
+
+	private Row convertRow(JsonNode node, RowTypeInfo info, boolean mapByName) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java]: This is due to a weird Jackson behavior. The index getter does not work.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+			csvSchema.getArrayElementSeparator(),
+			csvSchema.getQuoteChar(),
+			csvSchema.getEscapeChar(),
+			csvSchema.getNullValue());
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private void convertRow(ObjectNode reuse, Row row, RowTypeInfo rowTypeInfo) {
+		if (reuse == null) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java]: Good point :)

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+			return convertObjectArray((Object[]) obj, ((ObjectArrayTypeInfo) info).getComponentInfo());
+		} else if (info instanceof PrimitiveArrayTypeInfo &&
+				((PrimitiveArrayTypeInfo) info).getComponentType() == Types.BYTE) {
+			return container.binaryNode((byte[]) obj);
+		} else {
+			throw new RuntimeException("Unsupported type information '" + info + "' for object: " + obj);
+		}
+	}
+
+	private ArrayNode convertRow(Row row, RowTypeInfo rowTypeInfo) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java]: Must not necessarily be nested.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java]: Right, got it. :( That's unfortunate...

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+	public boolean isEndOfStream(Row nextElement) {
+		return false;
+	}
+
+	@Override
+	public TypeInformation<Row> getProducedType() {
+		return typeInfo;
+	}
+
+	public void setFieldDelimiter(char c) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java]: I agree.

File: flink-formats/flink-csv/pom.xml [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+			<artifactId>flink-table-common</artifactId>
+			<version>${project.version}</version>
+			<scope>provided</scope>
+		</dependency>
+
+		<!-- test dependencies -->
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-test-utils-junit</artifactId>
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-formats/flink-csv/pom.xml]: Very good point. Nobody noticed it so far.

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-formats/flink-csv/pom.xml]: The same applies to `flink-json`.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java [hunk 7]
```diff
@@ -169,10 +194,7 @@ public int hashCode() {
 
 	// --------------------------------------------------------------------------------------------
 
-	private void convertRow(ObjectNode reuse, Row row, RowTypeInfo rowTypeInfo) {
-		if (reuse == null) {
-			reuse = csvMapper.createObjectNode();
-		}
+	private void convertNestedRow(ObjectNode reuse, Row row, RowTypeInfo rowTypeInfo) {
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java]: Shouldn't this be just `convertRow`?

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java [hunk 8]
```diff
@@ -240,19 +260,23 @@ private JsonNode convert(ContainerNode<?> container, Object obj, TypeInformation
 		}
 	}
 
-	private ArrayNode convertRow(Row row, RowTypeInfo rowTypeInfo) {
+	private ArrayNode convertNestedRow(Row row, RowTypeInfo rowTypeInfo) {
 		final ArrayNode arrayNode = csvMapper.createArrayNode();
 		final TypeInformation[] types = rowTypeInfo.getFieldTypes();
+		if (row.getArity() != types.length) {
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java]: Extract to helper method? The same block is used twice.

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvRowDeSerializationSchemaTest.java [hunk 9]
```diff
@@ -90,20 +90,20 @@ public void testSerializeDeserialize() throws IOException {
...
-	public void testCustomizedProperties() throws IOException {
+	public void testSerializeDeserializeCustomizedProperties() throws IOException {
 
-		final Consumer<CsvRowSerializationSchema> serConfig = (serSchema) -> {
-			serSchema.setEscapeCharacter('*');
-			serSchema.setQuoteCharacter('\'');
-			serSchema.setArrayElementDelimiter(":");
-			serSchema.setFieldDelimiter(';');
+		final Consumer<CsvRowSerializationSchema.Builder> serConfig = (serSchemaBuilder) -> {
+			serSchemaBuilder.setEscapeCharacter('*');
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvRowDeSerializationSchemaTest.java]: Use the fluent API?

File: flink-formats/flink-csv/src/test/java/org/apache/flink/table/descriptors/CsvTest.java [hunk 10]
```diff
@@ -43,6 +43,20 @@
 		)}
 	);
 
+	private static final Descriptor DESCRIPTOR_1 = new Csv()
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/table/descriptors/CsvTest.java]: `DESCRIPTOR_1` -> `CUSTOM_DESCRIPTOR_WITH_MANUAL_SCHEMA`
`DESCRIPTOR_2` -> `DEFAULT_DESCRIPTOR_WITH_DERIVED_SCHEMA`

File: flink-formats/flink-csv/src/test/java/org/apache/flink/table/descriptors/CsvTest.java [hunk 11]
```diff
@@ -56,27 +70,17 @@ public void testMissingSchema() {
 	@Test(expected = ValidationException.class)
 	public void testDuplicateSchema() {
 		// we add an additional schema
-		addPropertyAndVerify(descriptors().get(1), "format.schema", "DDD");
+		addPropertyAndVerify(
+			descriptors().get(1),
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/table/descriptors/CsvTest.java]: Use `DEFAULT_DESCRIPTOR_WITH_DERIVED_SCHEMA`?

Comment (dawidwys) [reply flink-formats/flink-csv/src/test/java/org/apache/flink/table/descriptors/CsvTest.java]: The same above.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java (hunk 7 above)
Comment (twalthr) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java]: Sorry, my IDE was too eager.

**Comment by twalthr** [comment]
Thanks for the review @dawidwys. I forgot to add the JAR to the SQL Client e2e test. I will do this while merging and also quickly very the format manually in the CLI.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java [hunk 12]
```diff
@@ -209,91 +213,157 @@ public int hashCode() {
...
-			return Date.valueOf(node.asText());
-		} else if (info == Types.SQL_TIME) {
-			return Time.valueOf(node.asText());
-		} else if (info == Types.SQL_TIMESTAMP) {
-			return Timestamp.valueOf(node.asText());
+	private interface RuntimeConverter extends Serializable {
+		Object convert(JsonNode node);
+	}
+
+	private static RuntimeConverter createRowRuntimeConverter(
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java]: I would split this method into sth like:

	private static RuntimeConverter createRowRuntimeConverter(
			RowTypeInfo rowTypeInfo,
			boolean ignoreParseErrors,
			boolean isTopLevel) {
		final TypeInformation<?>[] fieldTypes = rowTypeInfo.getFieldTypes();
		final String[] fieldNames = rowTypeInfo.getFieldNames();

		final RuntimeConverter[] fieldConverters = createFieldConverters(ignoreParseErrors, fieldTypes);

		// Jackson only supports mapping by name in the first level
		return assembleRowConverter(ignoreParseErrors, isTopLevel, fieldNames, fieldConverters);
	}

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java [hunk 13]
```diff
@@ -209,91 +213,157 @@ public int hashCode() {
...
+		final TypeInformation<?>[] fieldTypes = rowTypeInfo.getFieldTypes();
+		final String[] fieldNames = rowTypeInfo.getFieldNames();
+
+		final RuntimeConverter[] fieldConverters = new RuntimeConverter[fieldTypes.length];
+		for (int i = 0; i < fieldTypes.length; i++) {
+			final TypeInformation<?> fieldType = fieldTypes[i];
+			final RuntimeConverter fieldConverter =
+				createNullableRuntimeConverter(fieldType, ignoreParseErrors);
+
+			fieldConverters[i] = (node) -> {
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java]: Why not just `fieldConverters[i] = createNullableRuntimeConverter(fieldType, ignoreParseErrors);` ?

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java [hunk 14]
```diff
@@ -194,91 +199,121 @@ public int hashCode() {
 
 	// --------------------------------------------------------------------------------------------
 
-	private void convertRow(ObjectNode reuse, Row row, RowTypeInfo rowTypeInfo) {
-		final TypeInformation[] types = rowTypeInfo.getFieldTypes();
+	private interface RuntimeConverter extends Serializable {
+		JsonNode convert(CsvMapper csvMapper, ContainerNode<?> container, Object obj);
+	}
+
+	private static RuntimeConverter createRowRuntimeConverter(RowTypeInfo rowTypeInfo, boolean isTopLevel) {
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java]: Split similarly as in `Deserialization` schema

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java [hunk 15]
```diff
@@ -71,9 +75,10 @@
 	private transient ObjectNode root;
 
 	private CsvRowSerializationSchema(
-			TypeInformation<Row> typeInfo,
+			RowTypeInfo typeInfo,
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java]: Like this change! ;)

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java [hunk 16]
```diff
@@ -209,91 +213,157 @@ public int hashCode() {
...
+						throw t;
+					}
+					return null;
+				}
+			};
+		}
+
+		// Jackson only supports mapping by name in the first level
+		if (isTopLevel) {
+			return (node) -> {
```
Comment (dawidwys) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java]: Could we extract common class for those branches? The only difference is how we extract children of `JsonNode`.

**Comment by dawidwys** [comment]
@flinkbot approve architecture

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java (hunk 13 above)
Comment (twalthr) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java]: We could simplify the logic a bit. I agree. I will put the checking into the method.

**Comment by twalthr** [comment]
Thanks for the review @dawidwys. I addressed your comments. I think we don't need an end-to-end test for everything. Esp. with the serialization test we should be good now. If you think we should have those nevertheless. Feel free to open an follow-up issue.

Will merge it now due to feature freeze.
@flinkbot approve quality

**Pull request closed without merging**